ТЗ 8 - Время и длительность

Natali-chiss opened the pull request.

Natali-chiss committed:
ТЗ 8 - Время и длительность

TyutterinYakov commented on .idea/vcs.xml:
Действительно ли этот файл требуется сохранять и отслеживать?) Хорошо было бы его удалить и занести в `.gitignore`

TyutterinYakov commented on src/com/yandex/tasktracker/model/Task.java:
```diff
@@ -68,13 +82,43 @@ public int hashCode() {
+    }
+
+    public LocalDateTime getEndTime() {
+        if (startTime != null) {
```
Также не забываем проверить значение продолжительности, чтобы избежать NPE

TyutterinYakov commented on src/com/yandex/tasktracker/model/Task.java:
```diff
@@ -1,19 +1,33 @@
         this.name = name;
         this.description = description;
         this.status = status;
+        this.duration = Duration.ofMinutes(0);
```
В качесте дефолтно более явно было бы использовать `null`, так как это лучше указывает на то, что продолжительность не задана)

TyutterinYakov commented on src/com/yandex/tasktracker/service/NotFoundException.java:
Для исключений хорошо было бы завести отдельный пакет

TyutterinYakov commented on src/com/yandex/tasktracker/service/ValidationException.java:
Для исключений хорошо было бы завести отдельный пакет

TyutterinYakov commented on src/com/yandex/tasktracker/service/InMemoryTaskManager.java:
```diff
@@ -172,55 +209,103 @@ public void removeEpic(int id) {
+        LocalDateTime currentEndTime;
+        Duration duration = Duration.ofMinutes(0);
+
         if (!epic.getSubtasksIds().isEmpty()) {
```
Если у эпика не будет подзадач - будет установлена продолжительность равна 0, а должно быть `null`

TyutterinYakov commented on src/com/yandex/tasktracker/service/InMemoryTaskManager.java:
```diff
@@ -153,16 +180,26 @@ public List<Subtask> getSubtasksList() {
-        tasks.remove(id);
+        final Task task = tasks.remove(id);
+        if (task.getStartTime() != null) {
+            prioritizedTasks.remove(tasks.get(id));
```
Удаление из отсортированного списка не произойдет, так как выше мы уже удалили задачу из мапы

TyutterinYakov commented on src/com/yandex/tasktracker/service/InMemoryTaskManager.java:
```diff
@@ -153,16 +180,26 @@ public List<Subtask> getSubtasksList() {
         final Epic epic = epics.remove(id);
         for (Integer subtaskId : epic.getSubtasksIds()) {
+            if (subtasks.get(subtaskId).getStartTime() != null) {
+                prioritizedTasks.remove(subtasks.get(subtaskId));
```
Может быть можно как-то записать без дополнительного обращения к мапе здесь и ниже?)

TyutterinYakov commented on src/com/yandex/tasktracker/service/InMemoryTaskManager.java:
```diff
@@ -109,31 +126,41 @@ public void clearEpicsList() {
     public void clearSubtasksList() {
         for (Integer id : subtasks.keySet()) {
```
Хорошо было бы итерироваться по значениям мапы - `.values()`, чтобы на каждой итерации цикла не производить обращение к ней)

TyutterinYakov commented on src/com/yandex/tasktracker/service/InMemoryTaskManager.java:
```diff
@@ -100,6 +116,7 @@ public void clearEpicsList() {
         }
         for (Integer id : subtasks.keySet()) {
```
Хорошо было бы итерироваться по значениям мапы - `.values()`, чтобы на каждой итерации цикла не производить обращение к ней)

TyutterinYakov commented on src/com/yandex/tasktracker/service/InMemoryTaskManager.java:
```diff
@@ -74,21 +86,25 @@ public void updateEpic(Epic epic) {
 
     @Override
     public void clearTasksList() {
         for (Integer id : tasks.keySet()) {
```
Хорошо было бы итерироваться по значениям мапы - `.values()`, чтобы на каждой итерации цикла не производить обращение к ней)

TyutterinYakov commented on src/com/yandex/tasktracker/service/InMemoryTaskManager.java:
```diff
@@ -74,21 +86,25 @@ public void updateEpic(Epic epic) {
 
     @Override
-    public void updateSubtask(Subtask subtask) {
+    public void updateSubtask(Subtask subtask) throws NotFoundException {
```
Указание непроверямых исключений в заголовках методов - не требуется, так как они не нуждаются в обязательной обработке

TyutterinYakov commented on src/com/yandex/tasktracker/service/InMemoryTaskManager.java:
```diff
@@ -172,55 +209,103 @@ public void removeEpic(int id) {
-                return;
-            } else {
-                status = Status.IN_PROGRESS;
+            if (t.getStartTime().isBefore(task.getStartTime()) && t.getEndTime().isAfter(task.getStartTime())) {
```
Учитываются не все варианты пересечений) Например, когда время задач полностью совпадает, когда время старта входит внутрь другой задачи и тд)

Можно поступить проще и отталкиваться не от тех ситуаций, когда все плохо и стоит выбрасывать исключение, а наоборот) Например, время окончания новой до либо равно времени старта существующей - это (!endTime.isAfter(existStartTime)). Так же следует добавить подобное условие для второго исхода. В обоих if при true нужно будет просто сделать continue; (пропустить итерацию). А, если оба условия будут false, после них можно выбросить исключение)

TyutterinYakov reviewed (commented):
Добрый вечер! В целом - хорошо) Отметил то, что следует исправить и то, что можно сделать лучше) Хорошего вечера! Если возникнут вопросы по ревью - пиши в пачку, буду рад помочь🌃

Natali-chiss committed:
Исправлены замечания по ревью:
1) Файл vcs.xml удалён и занесен в gitignore
2) Создана папка для исключений
3) Исправлен метод checkNoTimeConflict
4) Удалены излишние обращения к мапам
5) Длительность по умолчанию null

TyutterinYakov commented on src/com/yandex/tasktracker/service/InMemoryTaskManager.java:
```diff
@@ -172,55 +211,106 @@ public void removeEpic(int id) {
+    }
+
+    private void checkNoTimeConflict(Task task) {
+        for (Task t : getPrioritizedTasks()) {
```
Лучше называть переменные более полно, чтобы лишь по одному названию уже было понятно, что она хранит)

TyutterinYakov commented on src/com/yandex/tasktracker/service/InMemoryTaskManager.java:
```diff
@@ -172,55 +211,106 @@ public void removeEpic(int id) {
+                .filter(subtask -> subtask.getStatus() == Status.NEW)
+                .count();
+
+        if (subtasksList.isEmpty()) {
```
Можно изменить порядок таким образом:
```
        if (numberN == subtasksList.size()) {
            epic.setStatus(Status.NEW);
        } else if (numberD == subtasksList.size()) {
            epic.setStatus(Status.DONE);
        } else {
            epic.setStatus(Status.IN_PROGRESS);
        }
```
Таким образом мы не будем дублировать код)

TyutterinYakov reviewed (approved):
Добрый день! Поздравляю с успешной сдачей спринта :) Сил и терпения на этом сложном, но интересном пути! Успехов в обучении и прекрасного дня!🌅🎇

Natali-chiss committed:
Исправлены замечания по ревью:
1) Дублирование кода в calculateStatus убрано;
2) Переменные носят полное имя
